app/controllers/albums_controller.rb

```diff
@@ -0,0 +1,74 @@
+class AlbumsController < ApplicationController
+
+  def index
+    @albums = Album.order(ranked: :desc).limit(50)
+  end
+
+  def show
+    @album = findAlbum
+  end
+
+  def edit
+    @album = findAlbum
+
+    if @album == nil
+          render :file => 'public/404.html',
+              :status => :not_found
+    end
+  end
+
+  def update
+    @album = findAlbum
+
+    if @album.update(title: params[:album][:title], recorded_by: params[:album][:recorded_by], description: params[:album][:description])
+      redirect_to action: "show"
+    else
+      render 'edit'
+    end
+  end
+
+  def new
+    @album = Album.new
+  end
+
+  def create
+    @params = params
+    @album = Album.new
+    @album.title = params[:album][:title]
+    @album.recorded_by = params[:album][:recorded_by]
+    @album.description = params[:album][:description]
+
+    if(@album.save)
+      #saved successfully; go to index
+      redirect_to :action => "index"
+    else
+      #validation failed; show the "new" form again
+      render :action => :new
+    end
+  end
+
+  def destroy
+    @album = findAlbum
+
+    if @album.class == Album
+      @album.destroy
+      redirect_to albums_path
+    end
+
+  end
+
+  def upvote
+    Album.increment_counter(:ranked, params[:id])
+    redirect_to request.referer
+  end
+
+  private
+
+  def findAlbum
+    if Album.exists?(params[:id].to_i)
+      return Album.find(params[:id].to_i)
+    else
+      render status: 404
```

Comment:
Since this method is rendering a response (in the case that the requested album doesn't exist), we should actually use it as a controller filter by putting a `before_action :findAlbum` line at the top of this file.

Right now, if we request an album that doesn't exist it actually results in an error while trying to do the `render` call on this line. This is because we're actually rendering the _default template_ (e.g. `app/views/albums/show.html.erb` in the case of the `show` action) with this call.

We could change the call to match what we have on line 15. However, this code would still fail for the `update` action because of a subtle effect of the way in which it's being used with the line `@album = findAlbum`.

The best thing to do would be to make this a controller filter, use the `render` version from line 15 above, and change line 69 to say `@album = Album.find(params[:id])` instead of `return`.